[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50241 Re-enable no-op peepholes
* #50222 Add Post Freezing Optimizations, turn on by default in torch.jit.freeze
* #50221 Peephole Optimize out conv(x).dim(), which prevents BN fusion
* #50220 [JIT] Factor out peephole to own test file
* **#50075 [JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion**
* #50074 [JIT] Frozen Graph Conv-BN fusion

Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less).

Differential Revision: [D25856265](https://our.internmc.facebook.com/intern/diff/D25856265)

Commit:
Add Frozen Conv-> Add/Sub/Mul/Div fusion

[ghstack-poisoned]

Commit:
Update on "Add Frozen Conv-> Add/Sub/Mul/Div fusion"




[ghstack-poisoned]

Commit:
Update on "Add Frozen Conv-> Add/Sub/Mul/Div fusion"




[ghstack-poisoned]

Commit:
Update on "Add Frozen Conv-> Add/Sub/Mul/Div fusion"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -107,10 +109,268 @@ void FoldFrozenConvBatchnorm(Block* b) {
+
+        auto stack_out = runNodeIfInputsAreConstant(mul_or_div);
+        TORCH_INTERNAL_ASSERT(stack_out && stack_out->size() == 1);
+        Tensor fuse_bias = (*stack_out)[0].toTensor();
```
Nit: in some places it's `at::Tensor`, here it's just `Tensor`.

ZolotukhinM commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -107,10 +109,268 @@ void FoldFrozenConvBatchnorm(Block* b) {
+// constant tensor must satisfy the following:
+// - with resizing, broadcast to w/ weight/bias tensor shape
+// - broadcast to the conv output shape
+// It needs to be have a shape that can resize to weight/bias
```
Typo: "to be have".

ZolotukhinM commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -107,10 +109,268 @@ void FoldFrozenConvBatchnorm(Block* b) {
+// is they all contain a dim with value = channels-out. In the
+// conv output tensor, this is in the second dimension,
+// so the pointwise op tensor may have a second dimension of
+// value == chanels-out, but all the other dimensions have to be 1
```
Typo: chanels.

ZolotukhinM reviewed (approved):
Looks good, but I didn't dig deeply into the mathematical logic of the transformation.

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less). 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less). 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less). 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less). 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less). 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion"


Adds Conv - Add/Sub/Mul/Div fusion for frozen models. This helps cover models like torchvision maskrcnn, which use a hand-rolled batchnorm implementation: https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/ops/misc.py#L45. 

I haven't tested results yet but I would expect a somewhat similar speed up as conv-bn fusion (maybe a little less).

Differential Revision: [D25856265](https://our.internmc.facebook.com/intern/diff/D25856265)

[ghstack-poisoned]
